fix: handle dual task errors with read-only args

kriyanshii opened the pull request:
## Description
ray: handle dual task errors with read-only args

- avoid writing to user-defined args when building RayTaskError hybrids
- fall back to RayTaskError-only with warning if subclassing fails
- add regression test covering read-only args user exceptions

## Related issues
Fixes #59437

kriyanshii committed:
Signed-off-by: kriyanshii <kriyanshishah06@gmail.com>

ray: handle dual task errors with read-only args

- avoid writing to user-defined args when building RayTaskError hybrids
- fall back to RayTaskError-only with warning if subclassing fails
- add regression test covering read-only args user exceptions

codope commented on python/ray/tests/test_traceback.py:
```diff
@@ -367,6 +367,29 @@ def raise_custom_exception():
+        ray.get(raise_read_only_args.remote())
+
+    assert isinstance(exc_info.value, RayTaskError)
+    assert "boom" in str(exc_info.value)
```
Consider adding following assertions:
```
assert isinstance(exc_info.value, ReadOnlyArgsError)  # verify dual inheritance
assert exc_info.value.args == (exc_info.value.cause,)  # verify args property works
```

codope reviewed (commented):
Overall LGTM. Just a small suggestion for more thorough testing.

kriyanshii committed:
Signed-off-by: kriyanshii <kriyanshishah06@gmail.com>

[core][test] Added both assertions to test_task_error_with_read_only_args_property

assert isinstance(exc_info.value, ReadOnlyArgsError) to confirm dual inheritance
assert exc_info.value.args == (exc_info.value.cause,) to confirm the args shim works

kriyanshii committed:
Merge branch 'master' into allow-args

kriyanshii committed:
Merge branch 'master' into allow-args

kriyanshii committed:
Merge branch 'master' into allow-args

codope reviewed (approved):
LGTM! 
@edoakes @dayshah to merge

kriyanshii committed:
Merge branch 'master' into allow-args
